### PR TITLE
[SPARK-48510][2/2] Support UDAF `toColumn` API in Spark Connect

### PR DESCRIPTION
**Pull request by xupefei**
### What changes were proposed in this pull request?

This PR follows https://github.com/apache/spark/pull/46245 to add support `udaf.toColumn` API in Spark Connect.

Here we introduce a new Protobuf message, `proto.TypedAggregateExpression`, that includes a serialized UDF packet. On the server, we unpack it into an `Aggregator` object and generate a real `TypedAggregateExpression` instance with the encoder information passed along with the UDF.

### Why are the changes needed?

Because the `toColumn` API is not supported in the previous PR.

### Does this PR introduce _any_ user-facing change?

Yes, from now on users could create typed UDAF using `udaf.toColumn` API/.

### How was this patch tested?

New tests.

### Was this patch authored or co-authored using generative AI tooling?

Nope.

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/Aggregator.scala [hunk 1]
```diff
@@ -52,7 +56,7 @@ import org.apache.spark.sql.{Encoder, TypedColumn}
  * @since 4.0.0
  */
 @SerialVersionUID(2093413866369130093L)
-abstract class Aggregator[-IN, BUF, OUT] extends Serializable {
+abstract class Aggregator[-IN: TypeTag, BUF, OUT] extends Serializable {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/Aggregator.scala:59]: There are a few concerns here:
- How does this work with java implementations? How ugly is this going to be.
- This is a breaking change compared to the sql/core class. I am fine adding this. However I'd probably homogenise the implementations.

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 2]
```diff
@@ -48,6 +48,7 @@ message Expression {
     CommonInlineUserDefinedFunction common_inline_user_defined_function = 15;
     CallFunction call_function = 16;
     NamedArgumentExpression named_argument_expression = 17;
+    TypedAggregateExpression typed_aggregate_expression = 18;
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:51]: Why do we need a separate expression? It is just an invocation with a wild-card.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/Aggregator.scala (hunk 1 above)
Comment (xupefei) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/Aggregator.scala]: I reverted this change and switched to use reflection. The new method is hacky but should work in all cases.

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 2 above)
Comment (xupefei) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:51]: With this new expression it is very easy to identify this `toColumn` usage and transform it into a query plan which 1:1 matches a plan produced by the non-Connect version of Spark.

> It is just an invocation with a wild-card

Could you elaborate? Do you mean something like...
`new UserDefinedFunction(...).apply(col("*")).expr` then create a TypedColumn from it?
I tried this one but it seems more modification is needed to get rid of the Star.

**Comment by xupefei** [comment]
@HyukjinKwon Hey Hyukjin, could you review this PR?

**Comment by HyukjinKwon** [comment]
Will review. mind resolving conflicts?

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 3]
```diff
@@ -401,6 +402,11 @@ message JavaUDF {
   bool aggregate = 3;
 }
 
+message TypedAggregateExpression {
+  // (Required) The aggregate function object packed into bytes.
+  ScalarScalaUDF scalar_scala_udf = 1;
+}
+
```
Comment (HyukjinKwon) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:409]: I think this needs some review from @grundprinzip

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

**Comment by HyukjinKwon** [comment]
@xupefei mind resolving conflcits? I will merge this in

**Comment by xupefei** [comment]
> @xupefei mind resolving conflcits? I will merge this in

Done!

**Comment by HyukjinKwon** [comment]
Merged to master.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/UserDefinedFunctionE2ETestSuite.scala [hunk 4]
```diff
@@ -388,6 +378,66 @@ class UserDefinedFunctionE2ETestSuite extends QueryTest with RemoteSparkSession
       .head()
     assert(result == 135) // 45 + 90
   }
+
+  test("UDAF custom Aggregator - toColumn") {
```
Comment (dongjoon-hyun) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/UserDefinedFunctionE2ETestSuite.scala:382]: This seems to break Maven build.

```
- UDAF custom Aggregator - toColumn *** FAILED ***
  java.lang.IllegalArgumentException: Type tag defined in JavaMirror with jdk.internal.loader.ClassLoaders$AppClassLoader@4554617c of type class jdk.internal.loader.ClassLoaders$AppClassLoader with classpath [<unknown>] and parent being jdk.internal.loader.ClassLoaders$PlatformClassLoader@27d4a09 of type class jdk.internal.loader.ClassLoaders$PlatformClassLoader with classpath [<unknown>] and parent being primordial classloader with boot classpath [<unknown>] cannot be migrated to other mirrors.
```

**Comment by dongjoon-hyun** [comment]
Hi, @xupefei , @hvanhovell and @HyukjinKwon .

Could you double-check all Maven build status and recover them, please?
- https://github.com/apache/spark/actions/workflows/build_maven.yml (Java 17)
- https://github.com/apache/spark/actions/workflows/build_maven_java21.yml (Java 21)

Although I re-triggered Today's run first, I expect it to fail.
Screenshot 2024-07-15 at 10 37 19

**Comment by HyukjinKwon** [comment]
@xupefei mind taking a look ASAP? Seems like it broke JDK 17/21 builds

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/UserDefinedFunctionE2ETestSuite.scala [hunk 5]
```diff
@@ -388,6 +378,66 @@ class UserDefinedFunctionE2ETestSuite extends QueryTest with RemoteSparkSession
...
+    val session: SparkSession = spark
+    import session.implicits._
+    val aggCol = new CompleteUdafTestInputAggregator().toColumn
+    val ds = spark.range(10).withColumn("extra", col("id") * 2).as[UdafTestInput]
+
+    assert(ds.select(aggCol).head() == 135) // 45 + 90
+    assert(ds.agg(aggCol).head().getLong(0) == 135) // 45 + 90
+  }
+
+  test("UDAF custom Aggregator - multiple extends - toColumn") {
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/UserDefinedFunctionE2ETestSuite.scala:392]: https://github.com/apache/spark/actions/runs/9957362132/job/27509200696

image


Both tests 'UDAF custom Aggregator - toColumn' and 'UDAF custom Aggregator - multiple extends - toColumn' failed during Maven daily testing. Do you have time to fix them? @xupefei

Comment (HyukjinKwon) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/UserDefinedFunctionE2ETestSuite.scala:392]: We already landed a fix

**Pull request closed without merging**